Source/Generic/ECIODelegate.h

```diff
@@ -0,0 +1,101 @@
+//
+//  ECIODelegate.h
+//  ECLogging
+//
+//  Created by Mathieu Dutour on 18/10/2017.
+//  Copyright © 2017 Elegant Chaos. All rights reserved.
+//
+
+/**
+ Defines IO methods required by ECCommandLineEngine.
```

Comment:
I'd maybe change this description slightly, and not mention `ECCommandLineEngine`, since it's at a higher level and theoretically isn't known about by `ECLogging`.

Something like this:

---

Defines an abstract IO API for doing two things:
- reading settings
- outputting structured information

The settings are of a form that might be supplied in a dictionary, via a command line interface, or from some other key/value store.

The output of information is structured in the sense that it can be nested by opening and closing groups. The intention here is to allow a tool to produce human readable output but also to have enough context to optionally produce machine-readable output instead.

The protocol provides explicit methods for outputting formatted strings and errors, but also for outputting arbitrary "information" associated with a key. These can just be objects of any type.

A naive tool that supports this protocol can choose to log out all information in a flat way. By supporting keys and grouping, however, the API also allows a tool to create meaningful output that is more structured - eg JSON or XML.